userspace/libsinsp/sinsp_resource_utilization.cpp

```diff
@@ -0,0 +1,133 @@
+{
+	char filename[1024];
+	char line[512];
+	pid_t pid = getpid();
+
+	/* Path as seen from the calling process pid namespace. */
+	snprintf(filename, sizeof(filename), "/proc/%d/status", pid);
+	if(FILE* f = fopen(filename, "r"))
+	{
+		while(fgets(line, sizeof(line), f) != NULL)
+		{
+			if(strstr(line, "VmSize:") == line)
+			{
+				sscanf(line, "VmSize: %" PRIu32, &vsz);		/* memory size returned in kb */
+				vsz *= 1024;		/* convert to bytes */
+			} else if(strstr(line, "VmRSS:") == line)
+			{
+				sscanf(line, "VmRSS: %" PRIu32, &rss);		/* memory size returned in kb */
+				rss *= 1024;		/* convert to bytes */
+			}
+		}
+		fclose(f);
+	}
+
+    snprintf(filename, sizeof(filename), "/proc/%d/smaps_rollup", pid);
+	if(FILE* f = fopen(filename, "r"))
+	{
+		while(fgets(line, sizeof(line), f) != NULL)
+		{
+			if(strstr(line, "Pss:") == line)
+			{
+				sscanf(line, "Pss: %" PRIu32, &pss);		/* memory size returned in kb */
+				pss *= 1024;		/* convert to bytes */
+				break;
+			}
+		}
+		fclose(f);
+	}
+}
+
+void libsinsp::resource_utilization::get_cpu_usage(double &cpu_usage_perc, const double &self_pid_duration_sec)
+{
+
+	/* Number of clock ticks per second. */
+	double clocks_per_second = sysconf(_SC_CLK_TCK);
+
+	struct tms time;
+	if (times (&time) != (clock_t) -1)
+	{
+		/* utime is amount of processor time in user mode of calling process. Convert from clock_t to seconds. */
+		double user_sec = (double) time.tms_utime / clocks_per_second;
+		/* stime is amount of time the calling process has been scheduled in kernel mode. Convert from clock_t to seconds. */
+		double system_sec = (double) time.tms_stime / clocks_per_second;
+
+		/* CPU usage as percentage is computed by dividing the time the process uses the CPU by the elapsed time of the process. Compare to `ps` linux util. */
+		cpu_usage_perc = (user_sec + system_sec) / self_pid_duration_sec;
+		cpu_usage_perc *= 100.0;
+	}
+}
+
+void libsinsp::resource_utilization::get_container_memory_usage(uint32_t &memory_used)
+{
+    /* Path as seen from the calling process pid namespace.
+     * Hard-coded for "cloud-native" / Kubernetes support -> daemonset like deployment.
+     *
+     * In Kubernetes container_memory_working_set_bytes is the memory measure the OOM killer uses
+     * and values from /sys/fs/cgroup/memory/memory.usage_in_bytes are close enough
+     * and keep the approach adopted here simpler.
+     * -> carefully examined numbers from multiple sources in a real-life Kubernetes cluster.
+     *
+     * Please note that `kubectl top pod` numbers would reflect the sum of containers in a pod and
+     * typically libs clients (e.g. Falco) pods contain sidekick containers that use memory as well,
+     * therefore this metric accounts only for the container with the security monitoring agent running.
+    */
+    char filename[1024] = "/sys/fs/cgroup/memory/memory.usage_in_bytes";
```

Comment:
Above statement that this is close enough to Kubernetes `container_memory_working_set_bytes` that the OOM killer uses, see Kubernetes [metrics, metrics-server](https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/) is peer-review by infrastructure folks from my organization. We compared the numbers from Kubernetes prometheus endpoints with `memory.usage_in_bytes` after execing into a running container ...